### PR TITLE
Raise error upon request of backpropagation on device with finite shots

### PR DESCRIPTION
**Pull request by albi3ro**
Fixes Issue #1481 on the new QNode, as previous fix #1588 only solved the problem on the old QNode.

Contrary to the last bugfix on this problem, this fix raises an error instead of a warning.

```
dev = qml.device('default.qubit.autograd', shots=10, wires=1)

@qml.qnode(dev, interface='autograd', diff_method='backprop')
def circ(x):
    qml.RX(x, wires=0)
    return qml.expval(qml.PauliZ(0))
```
Now raises a `QuantumFunctionError`.

[sc-8003]

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -404,6 +404,9 @@
 
 <h3>Bug fixes</h3>
 
+* An error is raised during QNode creation if backpropagation is requested on a device with
+  finite-shots specified.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:408]:
```suggestion
  finite-shots specified.
  [(#2114)](https://github.com/PennyLaneAI/pennylane/pull/2114)
```

File: tests/test_qnode.py [hunk 2]
```diff
@@ -18,7 +18,7 @@
 from scipy.sparse import coo_matrix
 
 import pennylane as qml
-from pennylane import numpy as pnp
+from pennylane import QuantumFunctionError, numpy as pnp
```
Comment (albi3ro) [inline tests/test_qnode.py:21]:
```suggestion
from pennylane import numpy as pnp
```

I think my IDE automatically made that... oops

**Review by rmoyard — APPROVED** [review]
@albi3ro Thank you for changing that, it looks good to me!

**Comment by antalszava** [comment]
[sc-13322]

**Pull request merged**